File: src/test/java/no/nav/sosialhjelp/soknad/business/service/oppsummering/steg/inntektformue/SkattbarInntektTest.java
package no.nav.sosialhjelp.soknad.business.service.oppsummering.steg.inntektformue;

import no.nav.sbl.soknadsosialhjelp.soknad.JsonDriftsinformasjon;
import no.nav.sbl.soknadsosialhjelp.soknad.common.JsonKilde;
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.JsonOkonomi;
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.JsonOkonomiopplysninger;
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.opplysning.JsonOkonomiOpplysningUtbetaling;
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.opplysning.JsonOkonomibekreftelse;
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.opplysning.JsonOrganisasjon;
import no.nav.sosialhjelp.soknad.web.rest.ressurser.oppsummering.dto.SvarType;
import no.nav.sosialhjelp.soknad.web.rest.ressurser.oppsummering.dto.Type;
import org.junit.jupiter.api.Test;

import java.util.List;

import static java.util.Collections.singletonList;
import static no.nav.sbl.soknadsosialhjelp.json.SoknadJsonTyper.UTBETALING_SKATTEETATEN;
import static no.nav.sbl.soknadsosialhjelp.json.SoknadJsonTyper.UTBETALING_SKATTEETATEN_SAMTYKKE;
import static no.nav.sosialhjelp.soknad.business.service.oppsummering.steg.OppsummeringTestUtils.validateFeltMedSvar;
import static org.assertj.core.api.Assertions.assertThat;

class SkattbarInntektTest {

    private final SkattbarInntekt skattbarInntekt = new SkattbarInntekt();

    @Test
    void manglerSamtykke() {
        var okonomi = createOkonomi(false);
        var driftsinformasjon = new JsonDriftsinformasjon();

        var avsnitt = skattbarInntekt.getAvsnitt(okonomi, driftsinformasjon);

        assertThat(avsnitt.getSporsmal()).hasSize(1);
        assertThat(avsnitt.getSporsmal().get(0).getTittel()).isEqualTo("utbetalinger.inntekt.skattbar.mangler_samtykke");
    }

    @Test
    void feilMotSkatteetaten() {
        var okonomi = createOkonomi(true);
        var driftsinformasjon = new JsonDriftsinformasjon().withInntektFraSkatteetatenFeilet(true);

        var avsnitt = skattbarInntekt.getAvsnitt(okonomi, driftsinformasjon);

        assertThat(avsnitt.getSporsmal()).hasSize(1);
        assertThat(avsnitt.getSporsmal().get(0).getTittel()).isEqualTo("utbetalinger.skattbar.kontaktproblemer.oppsummering");
    }

    @Test
    void ingenSkattbareInntekter() {
        var okonomi = createOkonomi(true);
        var driftsinformasjon = new JsonDriftsinformasjon();

        var avsnitt = skattbarInntekt.getAvsnitt(okonomi, driftsinformasjon);

        assertThat(avsnitt.getSporsmal()).hasSize(2);

        var bekreftelseTidspunktSporsmal = avsnitt.getSporsmal().get(0);
        assertThat(bekreftelseTidspunktSporsmal.getTittel()).isEqualTo("utbetalinger.inntekt.skattbar.har_gitt_samtykke");
        assertThat(bekreftelseTidspunktSporsmal.getErUtfylt()).isTrue();
        assertThat(bekreftelseTidspunktSporsmal.getFelt()).hasSize(1);
        validateFeltMedSvar(bekreftelseTidspunktSporsmal.getFelt().get(0), Type.TEKST, SvarType.TIDSPUNKT, "2018-10-04T13:37:00.134Z");

        var inntekterSporsmal = avsnitt.getSporsmal().get(1);
        assertThat(inntekterSporsmal.getTittel()).isEqualTo("utbetalinger.inntekt.skattbar.inntekt.tittel");
        assertThat(inntekterSporsmal.getErUtfylt()).isTrue();
        assertThat(inntekterSporsmal.getFelt()).hasSize(1);
        validateFeltMedSvar(inntekterSporsmal.getFelt().get(0), Type.TEKST, SvarType.LOCALE_TEKST, "utbetalinger.inntekt.skattbar.ingen");
    }

    @Test
    void harEnSkattbarInntekt() {
        var okonomi = createOkonomi(true);
        okonomi.getOpplysninger()
                .setUtbetaling(
                        List.of(createUtbetaling("2020-01-01", "2020-02-01", 1234d, 123d))
                );
        var driftsinformasjon = new JsonDriftsinformasjon();

        var avsnitt = skattbarInntekt.getAvsnitt(okonomi, driftsinformasjon);

        assertThat(avsnitt.getSporsmal()).hasSize(2);

        var bekreftelseTidspunktSporsmal = avsnitt.getSporsmal().get(0);
        assertThat(bekreftelseTidspunktSporsmal.getTittel()).isEqualTo("utbetalinger.inntekt.skattbar.har_gitt_samtykke");
        assertThat(bekreftelseTidspunktSporsmal.getErUtfylt()).isTrue();
        assertThat(bekreftelseTidspunktSporsmal.getFelt()).hasSize(1);
        validateFeltMedSvar(bekreftelseTidspunktSporsmal.getFelt().get(0), Type.TEKST, SvarType.TIDSPUNKT, "2018-10-04T13:37:00.134Z");

        var inntekterSporsmal = avsnitt.getSporsmal().get(1);
        assertThat(inntekterSporsmal.getTittel()).isEqualTo("utbetalinger.inntekt.skattbar.inntekt.tittel");
        assertThat(inntekterSporsmal.getErUtfylt()).isTrue();
        assertThat(inntekterSporsmal.getFelt()).hasSize(1);
        var inntekt = inntekterSporsmal.getFelt().get(0);
        assertThat(inntekt.getType()).isEqualTo(Type.SYSTEMDATA_MAP);
        assertThat(inntekt.getLabelSvarMap()).hasSize(5);
        assertThat(inntekt.getLabelSvarMap().get("utbetalinger.utbetaling.arbeidsgivernavn.label").getValue()).isEqualTo("arbeidsgiver");
        assertThat(inntekt.getLabelSvarMap().get("utbetalinger.utbetaling.periodeFom.label").getValue()).isEqualTo("2020-01-01");
        assertThat(inntekt.getLabelSvarMap().get("utbetalinger.utbetaling.periodeTom.label").getValue()).isEqualTo("2020-02-01");
        assertThat(inntekt.getLabelSvarMap().get("utbetalinger.utbetaling.brutto.label").getValue()).isEqualTo("1234.0");
        assertThat(inntekt.getLabelSvarMap().get("utbetalinger.utbetaling.skattetrekk.label").getValue()).isEqualTo("123.0");
    }

    private JsonOkonomi createOkonomi(boolean harSamtykke) {
        return new JsonOkonomi()
                .withOpplysninger(
                        new JsonOkonomiopplysninger()
                                .withBekreftelse(
                                        singletonList(
                                                new JsonOkonomibekreftelse()
                                                        .withType(UTBETALING_SKATTEETATEN_SAMTYKKE)
                                                        .withVerdi(harSamtykke)
                                                        .withBekreftelsesDato("2018-10-04T13:37:00.134Z")
                                        )
                                )
                );
    }

    private JsonOkonomiOpplysningUtbetaling createUtbetaling(String fom, String tom, Double brutto, Double skattetrekk) {
        return new JsonOkonomiOpplysningUtbetaling()
                .withType(UTBETALING_SKATTEETATEN)
                .withKilde(JsonKilde.SYSTEM)
                .withOrganisasjon(new JsonOrganisasjon().withNavn("arbeidsgiver"))
                .withPeriodeFom(fom)
                .withPeriodeTom(tom)
                .withBrutto(brutto)
                .withSkattetrekk(skattetrekk);
    }
}

File: src/test/kotlin/no/nav/sosialhjelp/soknad/oppsummering/steg/inntektformue/SkattbarInntektTest.kt
package no.nav.sosialhjelp.soknad.oppsummering.steg.inntektformue

import no.nav.sbl.soknadsosialhjelp.json.SoknadJsonTyper
import no.nav.sbl.soknadsosialhjelp.soknad.JsonDriftsinformasjon
import no.nav.sbl.soknadsosialhjelp.soknad.common.JsonKilde
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.JsonOkonomi
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.JsonOkonomiopplysninger
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.opplysning.JsonOkonomiOpplysningUtbetaling
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.opplysning.JsonOkonomibekreftelse
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.opplysning.JsonOrganisasjon
import no.nav.sosialhjelp.soknad.oppsummering.dto.SvarType
import no.nav.sosialhjelp.soknad.oppsummering.dto.Type
import no.nav.sosialhjelp.soknad.oppsummering.steg.OppsummeringTestUtils.validateFeltMedSvar
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

internal class SkattbarInntektTest {
    private val skattbarInntekt = SkattbarInntekt()

    @Test
    fun manglerSamtykke() {
        val okonomi = createOkonomi(false)
        val driftsinformasjon = JsonDriftsinformasjon(false)

        val avsnitt = skattbarInntekt.getAvsnitt(okonomi, driftsinformasjon)
        assertThat(avsnitt.sporsmal).hasSize(1)
        assertThat(avsnitt.sporsmal[0].tittel).isEqualTo("utbetalinger.inntekt.skattbar.mangler_samtykke")
    }

    @Test
    fun feilMotSkatteetaten() {
        val okonomi = createOkonomi(true)
        val driftsinformasjon = JsonDriftsinformasjon(true)

        val avsnitt = skattbarInntekt.getAvsnitt(okonomi, driftsinformasjon)
        assertThat(avsnitt.sporsmal).hasSize(1)
        assertThat(avsnitt.sporsmal[0].tittel).isEqualTo("utbetalinger.skattbar.kontaktproblemer.oppsummering")
    }

    @Test
    fun ingenSkattbareInntekter() {
        val okonomi = createOkonomi(true)
        val driftsinformasjon = JsonDriftsinformasjon(false)

        val avsnitt = skattbarInntekt.getAvsnitt(okonomi, driftsinformasjon)
        assertThat(avsnitt.sporsmal).hasSize(2)

        val bekreftelseTidspunktSporsmal = avsnitt.sporsmal[0]
        assertThat(bekreftelseTidspunktSporsmal.tittel).isEqualTo("utbetalinger.inntekt.skattbar.har_gitt_samtykke")
        assertThat(bekreftelseTidspunktSporsmal.erUtfylt).isTrue
        assertThat(bekreftelseTidspunktSporsmal.felt).hasSize(1)
        validateFeltMedSvar(bekreftelseTidspunktSporsmal.felt!![0], Type.TEKST, SvarType.TIDSPUNKT, "2018-10-04T13:37:00.134Z")

        val inntekterSporsmal = avsnitt.sporsmal[1]
        assertThat(inntekterSporsmal.tittel).isEqualTo("utbetalinger.inntekt.skattbar.inntekt.tittel")
        assertThat(inntekterSporsmal.erUtfylt).isTrue
        assertThat(inntekterSporsmal.felt).hasSize(1)
        validateFeltMedSvar(inntekterSporsmal.felt!![0], Type.TEKST, SvarType.LOCALE_TEKST, "utbetalinger.inntekt.skattbar.ingen")
    }

    @Test
    fun harEnSkattbarInntekt() {
        val okonomi = createOkonomi(true).copy(opplysninger = createOkonomi(true).opplysninger.copy(utbetaling = listOf(createUtbetaling("2020-01-01", "2020-02-01", 1234.0, 123.0))))
        val driftsinformasjon = JsonDriftsinformasjon(false)

        val avsnitt = skattbarInntekt.getAvsnitt(okonomi, driftsinformasjon)
        assertThat(avsnitt.sporsmal).hasSize(2)

        val bekreftelseTidspunktSporsmal = avsnitt.sporsmal[0]
        assertThat(bekreftelseTidspunktSporsmal.tittel).isEqualTo("utbetalinger.inntekt.skattbar.har_gitt_samtykke")
        assertThat(bekreftelseTidspunktSporsmal.erUtfylt).isTrue
        assertThat(bekreftelseTidspunktSporsmal.felt).hasSize(1)
        validateFeltMedSvar(bekreftelseTidspunktSporsmal.felt!![0], Type.TEKST, SvarType.TIDSPUNKT, "2018-10-04T13:37:00.134Z")

        val inntekterSporsmal = avsnitt.sporsmal[1]
        assertThat(inntekterSporsmal.tittel).isEqualTo("utbetalinger.inntekt.skattbar.inntekt.tittel")
        assertThat(inntekterSporsmal.erUtfylt).isTrue
        assertThat(inntekterSporsmal.felt).hasSize(1)

        val inntekt = inntekterSporsmal.felt!![0]
        assertThat(inntekt.type).isEqualTo(Type.SYSTEMDATA_MAP)
        assertThat(inntekt.labelSvarMap).hasSize(5)
        assertThat(inntekt.labelSvarMap!!["utbetalinger.utbetaling.arbeidsgivernavn.label"]!!.value).isEqualTo("arbeidsgiver")
        assertThat(inntekt.labelSvarMap["utbetalinger.utbetaling.periodeFom.label"]!!.value).isEqualTo("2020-01-01")
        assertThat(inntekt.labelSvarMap["utbetalinger.utbetaling.periodeTom.label"]!!.value).isEqualTo("2020-02-01")
        assertThat(inntekt.labelSvarMap["utbetalinger.utbetaling.brutto.label"]!!.value).isEqualTo("1234.0")
        assertThat(inntekt.labelSvarMap["utbetalinger.utbetaling.skattetrekk.label"]!!.value).isEqualTo("123.0")
    }

    private fun createOkonomi(harSamtykke: Boolean): JsonOkonomi {
        return JsonOkonomi(JsonOkonomiopplysninger(emptyList(), listOf(JsonOkonomibekreftelse(JsonKilde.BRUKER, SoknadJsonTyper.UTBETALING_SKATTEETATEN_SAMTYKKE, "", harSamtykke, "2018-10-04T13:37:00.134Z")), null, emptyList(), null), null)
    }

    private fun createUtbetaling(
        fom: String,
        tom: String,
        brutto: Double,
        skattetrekk: Double,
    ): JsonOkonomiOpplysningUtbetaling {
        return JsonOkonomiOpplysningUtbetaling(JsonKilde.SYSTEM, SoknadJsonTyper.UTBETALING_SKATTEETATEN, "", false, brutto = brutto, skattetrekk = skattetrekk, periodeFom = fom, periodeTom = tom, organisasjon = JsonOrganisasjon("arbeidsgiver", ""))
    }
}
